File: src/parser.rs
use crate::console;
use crate::lexer::{Range, Token, TokenData};
use crate::prelude::OptExt;
use crate::project::project::{Operator, ProjectType};
use crate::project::unchecked_project::*;
use std::slice::Iter;

/// An iterator over input tokens for a parser.
#[derive(Clone)]
struct ParserInput<'a>(Iter<'a, Token>);

impl<'a> Iterator for ParserInput<'a> {
    type Item = &'a Token;

    fn next(&mut self) -> Option<&'a Token> {
        self.0.next()
    }
}

/// The current state for a parser. Allows setting and restoring a backup point, for if a lookahead
/// of 1 is not enough.
#[derive(Clone)]
struct ParserState<'a> {
    input: ParserInput<'a>,
    current_token: &'a Token,
    next_token: &'a Token,
    prev_token: &'a Token,
}

impl<'a> ParserState<'a> {
    fn new(tokens: &'a Vec<Token>) -> Self {
        let mut input = ParserInput(tokens.iter());
        if let Some(tkn) = input.next() {
            let next_tkn = input.next().unwrap_or(tkn);
            return ParserState {
                current_token: tkn,
                next_token: next_tkn,
                prev_token: tkn,
                input: input.clone(),
            };
        }

        console::return_with_error("Error parsing, no input.")
    }

    /// Returns a backup copy of the current state.
    fn backup(self: &mut Self) -> Self {
        self.clone()
    }

    /// Restores the current state to the provided backup.
    fn restore(self: &mut Self, backup: &Self) {
        self.clone_from(backup);
    }

    /// Moves to the next token in the input.
    fn next(self: &mut Self) {
        self.prev_token = self.current_token;
        self.current_token = self.next_token;
        self.next_token = self.input.next().unwrap_or(self.next_token);
    }

    /// Can be used to debug the current state while parsing.
    /// Displays the specified message, and the current token, including its position.
    #[allow(dead_code)]
    fn dbg(self: &mut Self, msg: &str) {
        println!("{}: {}", msg, self.current_token.to_string());
    }
}

/// Performs a check on the current token, and if it fails, displays a message that parsing the
/// entity with the provided label failed, and exits with code 1.
/// If it succeeds, moves to the next token.
fn check_and_next<'a>(state: &'a mut ParserState, f: fn(&Token) -> bool, label: &str) {
    let t = state.current_token;
    let msg = format!("{}: Error parsing {}, unexpected {}.", t.range.to_full_string(), label, t.data.to_string(true),);
    console::test_condition(f(t), msg.as_str());
    state.next();
}

/// Tries to consume a token, if the check on the token returns Some. Returns None otherwise.
/// Doesn't consume a token if the check fails.
fn try_consume<'a, T>(state: &'a mut ParserState, f: fn(&Token) -> Option<T>) -> Option<T> {
    match f(state.current_token) {
        Some(v) => {
            state.next();
            Some(v)
        }
        None => None,
    }
}

/// Consumes the current token, if the check on the token returns Some. If the check returns None,
/// displays an error message that parsing the entity with the provided label failed, and exits
/// with error code 1.
fn consume<'a, T>(state: &'a mut ParserState, f: fn(&Token) -> Option<T>, label: &str) -> T {
    let t = state.current_token;
    state.next();
    match f(t) {
        Some(v) => v,
        None => {
            let msg =
                format!("{}: Error parsing {}, unexpected {}.", t.range.to_string(), label, t.data.to_string(true));
            console::return_with_error(msg.as_str())
        }
    }
}

/// Consumes an end of line token.
fn consume_eol<'a>(state: &'a mut ParserState, label: &str) {
    check_and_next(state, |t| t.data == TokenData::EndOfLineToken(), format!("{} eol", label).as_str());
}

/// Consumes one or more end of line tokens. All but the first may be prefixed by indentation
/// tokens of any level.
fn consume_eols<'a>(state: &'a mut ParserState, label: &str) {
    fn try_consume_empty_line<'a>(state: &'a mut ParserState) -> bool {
        let indented = state.current_token.data.is_indentation(); // Consume any indentation first.

        // Then check if there is an end of line token.
        let t = if indented { state.next_token } else { state.current_token };
        if t.data != TokenData::EndOfLineToken() {
            return false;
        }

        if indented {
            state.next();
        }
        state.next();
        true
    }

    consume_eol(state, label); // There has to be at least one end of line.
    while try_consume_empty_line(state) {} // Then any number of empty lines.
}

/// Parses a project type.
fn parse_project_type<'a>(state: &'a mut ParserState) -> ProjectType {
    let start_range = &state.current_token.range;
    check_and_next(state, |t: &Token| t.data == TokenData::KeywordToken("Executable".to_string()), "project type");
    check_and_next(state, |t: &Token| t.data == TokenData::SymbolToken(":".to_string()), "project type");
    let name = consume(state, |t: &Token| t.data.try_get_identifier(), "project name");
    let end_range = &state.prev_token.range;
    consume_eols(state, "project type");
    ProjectType::Executable(Range::from_ranges(start_range, end_range), name)
}

/// Parses a separator between the project type definition and the program.
fn parse_separator<'a>(state: &'a mut ParserState) {
    check_and_next(state, |t: &Token| t.data == TokenData::SeparatorToken(), "separator");
    consume_eols(state, "separator");
}

/// Tries to pa rse a print statement, will return None if the first keyword is not matched and
/// fail if anything later fails.
fn try_parse_print_stmt<'a>(state: &'a mut ParserState) -> Option<(UncheckedStatement, bool)> {
    let start_token = &state.current_token.range;
    if state.current_token.data != TokenData::KeywordToken("print".to_string()) {
        return None;
    }

    state.next();

    let id_range = &state.current_token.range;
    let str_lit = try_consume(state, |t| t.data.try_get_string_literal())
        .map(|t| UncheckedStringLiteral::UStringLiteral(id_range.clone(), t));
    let expression = try_parse_expression(state);

    match (str_lit, expression) {
        (Some(sl), _) => Some((UncheckedStatement::UPrintStr(Range::from_ranges(start_token, id_range), sl), true)),
        (_, Some(expr)) => {
            Some((UncheckedStatement::UPrintExpr(Range::from_ranges(start_token, id_range), expr), true))
        }
        _ => {
            let msg = format!(
                "Error parsing a print statement, expected a {} or {}, but got {}.",
                TokenData::StringLiteralToken("".to_string()).to_string(false),
                TokenData::IdentifierToken("".to_string()).to_string(false),
                state.current_token.data.to_string(true)
            );
            console::return_with_error(msg.as_str())
        }
    }
}

/// Parses a (positive or negative) number.
fn try_parse_number<'a>(state: &'a mut ParserState) -> Option<i64> {
    let backup = state.backup();
    let is_negative = state.current_token.data == TokenData::SymbolToken("-".to_string());
    if is_negative {
        state.next();
    }

    try_consume(state, |t| t.data.try_get_number()).map(|n| if is_negative { -1 * n } else { n }).or_do(|| {
        state.restore(&backup);
    })
}

/// Tries to parse an operator.
fn try_parse_operator<'a>(state: &'a mut ParserState) -> Option<Operator> {
    try_consume(state, |c| c.data.try_get_symbol("+")).map(|_| Operator::Add(state.prev_token.range.clone())).or_else(
        || try_consume(state, |c| c.data.try_get_symbol("-")).map(|_| Operator::Sub(state.prev_token.range.clone())),
    )
}

/// Tries to parse an expression.
fn try_parse_expression<'a>(state: &'a mut ParserState) -> Option<UncheckedExpression> {
    /// Tries to parse a binary expression.
    fn try_parse_binary<'a>(state: &'a mut ParserState) -> Option<UncheckedExpression> {
        let backup = state.backup();
        try_parse_int_literal(state)
            .or_else(|| try_parse_variable(state))
            .bind(|l| {
                try_parse_operator(state).bind(|op| {
                    try_parse_expression(state).map(|ex| {
                        UncheckedExpression::UBinary(
                            Range::from_ranges(&backup.current_token.range.clone(), &state.current_token.range.clone()),
                            op,
                            Box::new(l),
                            Box::new(ex),
                        )
                    })
                })
            })
            .or_do(|| state.restore(&backup))
    }

    fn try_parse_variable<'a>(state: &'a mut ParserState) -> Option<UncheckedExpression> {
        let var = try_consume(state, |t| t.data.try_get_identifier());
        var.map(|name| {
            let expr = try_parse_expression(state);
            UncheckedExpression::UVariable(
                state.prev_token.range.clone(),
                UncheckedVariable {
                    range: state.prev_token.range.clone(),
                    name,
                },
                expr.map(Box::new),
            )
        })
    }

    fn try_parse_int_literal<'a>(state: &'a mut ParserState) -> Option<UncheckedExpression> {
        try_parse_number(state).map(|n| UncheckedExpression::UIntLiteral(state.prev_token.range.clone(), n))
    }

    // All of the sub parsers reset state, so no need to do it here.
    try_parse_binary(state).or_else(|| try_parse_int_literal(state)).or_else(|| try_parse_variable(state))
}

/// Tries to parse an assignment, which can be either directly an expression, or a block of
/// statements. Will return none if parsing the expression failed, or the block start was not
/// matched (end of line). Block statements will be parsed as long as the indentation is correct.
fn try_parse_assignment<'a>(state: &'a mut ParserState, indent: usize) -> Option<(UncheckedAssignment, bool)> {
    if state.current_token.data == TokenData::EndOfLineToken() {
        consume_eol(state, "block assignment");
        let range_start = &state.current_token.range;

        let mut stmts = Vec::<UncheckedStatement>::new();
        while let Some(stmt) = try_parse_statement(state, indent + 1) {
            stmts.push(stmt);
        }

        Some((
            UncheckedAssignment::UBlockAssignment(Range::from_ranges(range_start, &state.prev_token.range), stmts),
            false,
        ))
    } else {
        try_parse_expression(state).map(|e| (UncheckedAssignment::UExprAssignment(e.get_range(), e), true))
    }
}

/// Tries to parse an assignment statement, will return None if the first keyword is not matched and fail if
/// anything later fails.
fn try_parse_assignment_stmt<'a>(state: &'a mut ParserState, indent: usize) -> Option<(UncheckedStatement, bool)> {
    let range_start = &state.current_token.range;
    if state.current_token.data != TokenData::KeywordToken("let".to_string()) {
        return None;
    }

    state.next();

    let name_range = &state.current_token.range;
    let name = consume(state, |t| t.data.try_get_identifier(), "assignment variable");

    let param_name_range = &state.current_token.range;
    let param_name = try_consume(state, |t| t.data.try_get_identifier());

    check_and_next(state, |t| t.data == TokenData::SymbolToken("=".to_string()), "assignment");

    let (assmt, eols) = try_parse_assignment(state, indent)
        .assert_some(|| console::return_with_error("Failed to parse an assignment."));
    Some((
        UncheckedStatement::UAssignment(
            Range::from_ranges(range_start, &state.prev_token.range),
            UncheckedVariable {
                range: name_range.clone(),
                name,
            },
            param_name.map(|n| UncheckedVariable {
                range: param_name_range.clone(),
                name: n,
            }),
            assmt,
        ),
        eols,
    ))
}

/// Tries to parse a return statement, will return None if the first token is not a return keyword,
/// and fail if anything later fails.
fn try_parse_return_stmt<'a>(state: &'a mut ParserState) -> Option<(UncheckedStatement, bool)> {
    let range_start = &state.current_token.range;
    if state.current_token.data != TokenData::KeywordToken("return".to_string()) {
        return None;
    }

    state.next();

    try_parse_expression(state)
        .map(|e| Some((UncheckedStatement::UReturn(Range::from_ranges(range_start, &state.prev_token.range), e), true)))
        .assert_some(|| console::return_with_error("Failed to parse a return expression"))
}

/// Checks that the next token is indented to the specified indent level.
/// For an indent level of 0, it is checked that the next token is not an indent token, and no
/// tokens are consumed.
/// If the next token is the desired indent token (greater than 0), it is consumed, otherwise
/// nothing is consumed.
fn check_indent<'a>(state: &'a mut ParserState, indent: usize) -> bool {
    match indent {
        0 => !state.current_token.data.is_indentation(),
        _ => {
            if state.current_token.data == TokenData::IndentationToken(indent) {
                state.next();
                true
            } else {
                false
            }
        }
    }
}

/// Tries to parse a statement. This pa rser first checks whether the next token has the correct
/// indentation, then applies onfe of the parsers for the specific statements. Consumes no tokens
/// if the indentation is incorrect, or no statement was parsed.
fn try_parse_statement<'a>(state: &'a mut ParserState, indent: usize) -> Option<UncheckedStatement> {
    let backup = state.backup();

    if !check_indent(state, indent) {
        return None;
    }

    try_parse_print_stmt(state)
        .or_else(|| try_parse_assignment_stmt(state, indent))
        .or_else(|| try_parse_return_stmt(state))
        .or_do(|| {
            state.restore(&backup);
        })
        .map(|(s, eols)| {
            if eols {
                consume_eols(state, "statement");
            }
            s
        })
}

/// Parses a program.
fn parse_program<'a>(state: &'a mut ParserState) -> UncheckedProgram {
    let start_range = &state.current_token.range;
    let mut stmts: Vec<UncheckedStatement> = Vec::new();
    let mut stmt_opt = try_parse_statement(state, 0);
    while let Some(stmt) = stmt_opt {
        stmts.push(stmt);
        // try_parse_top_level_statement already consumes end of lines.
        stmt_opt = try_parse_statement(state, 0);
    }

    UncheckedProgram {
        range: Range::from_ranges(start_range, &state.prev_token.range),
        stmts,
    }
}

/// Parses a project from a list of tokens.
/// The parsing functions in this parser can fail in two ways:
/// Either they return an Option, in which case there is a chance that another parser mayh be used
/// instead, or they directly return the parsed value, and failure means that parsing failed. Any
/// parser that returns None is responsible for ensuring that the state is reset to before it
/// performed any actions. In case of failure, the parsing function will print the error and exit
/// the program.
/// Note that a parser that returns an Option can still fail if it has decided that what it is
/// parsing is indeed what it is supposed to be, but the input later down is not correct.
pub fn parse_project<'a>(input: Vec<Token>) -> UncheckedProject {
    let mut state = ParserState::new(&input);

    let project_type = parse_project_type(&mut state);
    parse_separator(&mut state);
    let program = parse_program(&mut state);
    check_and_next(&mut state, |t: &Token| t.data == TokenData::EndOfInputToken(), "project");

    UncheckedProject { project_type, program }
}

File: src/Parser.hs
module Parser (Parser, projectParser, run) where

import Console (Formattable (formatBare))
import Control.Applicative (Alternative (empty, many, (<|>)), optional)
import Control.Monad.Loops (whileJust, whileM_)
import Control.Monad.Trans.Class (MonadTrans (lift))
import Control.Monad.Trans.Except (ExceptT, runExceptT, throwE)
import Control.Monad.Trans.State (State)
import qualified Control.Monad.Trans.State.Lazy as ST (evalState, get, gets, modify, put)
import Data.Bifunctor (Bifunctor (first, second))
import Data.Char (isLower)
import Data.List (uncons)
import qualified Data.List as List (filter, intercalate, uncons)
import qualified Data.Maybe as Maybe (isNothing)
import Data.Text (Text)
import qualified Data.Text as T
import Lexer
  ( Range,
    Token (..),
    TokenData (..),
    ignoreToken,
    isEol,
    isIndentation,
    isSeparator,
    rangeFromRanges,
    tryGetIdentifier,
    tryGetNumber,
    tryGetStringLiteral,
    tryGetSymbol,
  )
import Numeric.Natural (Natural)
import Project
import Text.Printf (printf)

{- ParserState -}

data ParserState = ParserState {input :: [Token], curTkn :: Token, prevTkn :: Token}

createParserState :: [Token] -> Either String ParserState
createParserState tokens =
  case List.uncons tokens of
    Just (cur, tail) -> Right $ ParserState {input = tail, prevTkn = cur, curTkn = cur}
    _ -> Left "Error parsing, no input."

-- | Moves to the next token in the input.
--   If the last token has been reached, this token will be moved into nxtTkn, curTKn and then prevTkn.
nextToken :: ParserState -> ParserState
nextToken state@ParserState {input, curTkn} =
  case List.uncons input of
    Just (next', input') -> state {prevTkn = curTkn, curTkn = next', input = input'}
    Nothing -> state {prevTkn = curTkn}

{- Parser types -}

data ParseResult a = Success a | Failure Text | Ignore deriving (Show)

instance Functor ParseResult where
  fmap f (Success a) = Success $ f a
  fmap _ (Failure err) = Failure err
  fmap _ Ignore = Ignore

instance Applicative ParseResult where
  pure = Success
  (<*>) (Success f) a = fmap f a
  (<*>) (Failure err1) (Failure err2) = Failure (T.intercalate "\n" [err1, err2])
  (<*>) (Failure err) _ = Failure err
  (<*>) _ (Failure err) = Failure err
  (<*>) _ _ = Ignore

instance Alternative ParseResult where
  empty = Ignore
  (<|>) (Success a) _ = Success a
  (<|>) (Failure err) _ = Failure err
  (<|>) Ignore b = b

instance Monad ParseResult where
  (>>=) (Success a) f = f a
  (>>=) (Failure err) f = Failure err
  (>>=) Ignore f = Ignore

-- | Converts a ParseResult to an Either. the Failure and Ignore cases are converted to Left.
parseResultToEither :: ParseResult a -> Either String a
parseResultToEither (Success a) = Right a
parseResultToEither (Failure err) = Left $ T.unpack err
parseResultToEither Ignore = Left "Parser produced no result."

newtype Parser s a = Parser {runParser :: s -> ParseResult (a, s)}

deriving instance Functor (Parser s)

instance Applicative (Parser s) where
  pure a = Parser $ pure . (a,)
  (<*>) (Parser f) (Parser a) = Parser $ \s ->
    case f s of
      Ignore -> Ignore
      Failure err -> Failure err
      Success (f', s') -> first f' <$> a s'

instance Alternative (Parser s) where
  empty = Parser $ const Ignore
  (<|>) (Parser a) (Parser b) = Parser $ \s -> a s <|> b s

instance Monad (Parser s) where
  (>>=) (Parser a) f = Parser $ \s ->
    case a s of
      Ignore -> Ignore
      Failure err -> Failure err
      Success (a', s') -> runParser (f a') s'

{- Default parsers -}

-- | A parser that always fails with the specified error.
pfail :: String -> Parser s a
pfail err = Parser $ const $ Failure $ T.pack err

-- | A parser that returns the current state.
get :: Parser s s
get = Parser $ \s -> pure (s, s)

-- | A parser that returns the specifed mapping over the current state.
--   Can be useful to get a specific field.
gets :: (s -> a) -> Parser s a
gets f = f <$> get

{- Application specific parsers -}

type Parser' = Parser ParserState

-- Alias for shorter for matting.
fb :: Formattable a => a -> String
fb = formatBare

-- | Runs a parser on the specified list of tokens.
run :: Parser' a -> [Token] -> Either String a
run (Parser p) input =
  fst <$> (parseResultToEither . p =<< createParserState (List.filter (not . ignoreToken . tData) input))

-- | Returns the next token from the input. Will be an EndOfInputToken every time once the end of the input has been
--   reached.
pToken :: Parser' Token
pToken = Parser $ \s ->
  let s' = nextToken s
   in pure (curTkn s, s')

-- | A parser that succeeds if the exact specified token is consumed and fails otherwise.
consumeExact :: TokenData -> String -> Parser' TokenData
consumeExact toConsume label = do
  tkn <- pToken
  if tData tkn == toConsume
    then pure $ tData tkn
    else pfail $ printf "%s Parsing %s failed, expected %s, but got %s." (fb $ tokenRange tkn) label (fb toConsume) (fb $ tData tkn)

-- | A parser that succeeds if the mapping on the consumed token returns Just, and fails otherwise.
consumeJust :: (TokenData -> Maybe a) -> String -> Parser' a
consumeJust f label = fst <$> consumeJust' f label

-- | consumeJust but also returns the token's range.
consumeJust' :: (TokenData -> Maybe a) -> String -> Parser' (a, Range)
consumeJust' f label = do
  tkn <- pToken
  case f $ tData tkn of
    Just a -> pure (a, tokenRange tkn)
    Nothing -> pfail $ printf "%s Parsing %s failed, unexpected %s." (fb $ tokenRange tkn) label (fb $ tData tkn)

-- | A parser that succeeds if the check on the consumed token returns True, and fails otherwise.
consumeIf :: (TokenData -> Bool) -> String -> Parser' ()
consumeIf f = consumeJust (\x -> if f x then Just () else Nothing)

-- | A parser that succeeds if the exact specified token is consumed and returns Ignore otherwise.
tryConsumeExact :: TokenData -> Parser' ()
tryConsumeExact toConsume = tryConsumeIf (== toConsume)

-- | A parser that succeeds if the mapping on the consumed token returns Just, and returns Ignore otherwise.
tryConsumeJust :: (TokenData -> Maybe a) -> Parser' a
tryConsumeJust f = fst <$> tryConsumeJust' f

-- | tryConsumeJust but also returns the token's range.
tryConsumeJust' :: (TokenData -> Maybe a) -> Parser' (a, Range)
tryConsumeJust' f = do
  tkn <- pToken
  case f $ tData tkn of
    Nothing -> empty
    Just a -> pure (a, tokenRange tkn)

-- | A parser that succeeds if the check on the consumed token returns True, and returns Ignore otherwise.
tryConsumeIf :: (TokenData -> Bool) -> Parser' ()
tryConsumeIf f = tryConsumeJust (\x -> if f x then Just () else Nothing)

-- | A parser that parses exactly one end of line token and fails otherwise.
eolParser :: String -> Parser' ()
eolParser label = consumeIf isEol $ printf "%a end of line" label

{- Parsers for domain types -}

-- | A parser that parses one end of line, and then any number of empty lines, optionally preceded by whitespace.
--   If the first end of line fails, the parser fails. Any other empty lines are optional.
eolsParser :: String -> Parser' ()
eolsParser label = do
  eolParser label
  () <$ many emptyLineParser
  where
    -- Parses an empty line, optionally with an indentation token in it. Returns Ignore if it was not matched.
    emptyLineParser = do
      tkn <- pToken
      case tData tkn of
        IndentationToken _ -> tryConsumeIf isEol
        EndOfLineToken _ -> pure ()
        _ -> empty

-- | A parser for a project type.
projectTypeParser :: Parser' ProjectType
projectTypeParser = do
  start <- gets (tokenRange . curTkn)
  consumeExact (KeywordToken "Executable") "project type"
  consumeExact (SymbolToken ":") "project type"
  name <- consumeJust tryGetIdentifier "project name"

  end <- gets (tokenRange . prevTkn)
  eolsParser "project type"
  pure $ Executable (rangeFromRanges start end) name

-- | A parser that consumes the specified indentation token, and returns Ignore if this was not possible.
--   An indentation of 0 means that the next token must not be an indentation token. Otherwise, a matching indentation
--   token is consumed.
indentParserM :: Natural -> Parser' ()
indentParserM 0 = do
  t <- gets (tData . curTkn)
  case t of
    IndentationToken _ -> empty
    _ -> pure ()
indentParserM indent = tryConsumeExact (IndentationToken indent)

-- | A parser  for an assignment, which can be either directly an expression, or a block of statements.
--   Will return Ignore if parsing the expression failed, or the block start was not matched (end of line).
--   Block statements will be parsed as long as the indentation is correct.
assignmentParserM :: Natural -> Parser' (UncheckedAssignment, Bool)
assignmentParserM indent = do
  startTkn <- gets curTkn
  if isEol $ tData startTkn
    then do
      eolParser "block assignment"
      start <- gets (tokenRange . curTkn)
      stmts <- many $ statementParserM (indent + 1)
      end <- gets (tokenRange . prevTkn)
      pure (UBlockAssignment (rangeFromRanges start end) stmts, False)
    else (\e -> (UExprAssignment (expressionRange e) e, True)) <$> expressionParserM

-- | A parser for an operator, returns Ignore if an operator could not be parsed.
operatorParserM :: Parser' Operator
operatorParserM = tryOp "+" Add <|> tryOp "-" Sub
  where
    tryOp sym typ = typ . snd <$> tryConsumeJust' (tryGetSymbol sym)

-- | A parser for a number, returns Ignore if a number cannot be fully parsed.
numberParserM :: Parser' (Int, Range)
numberParserM = do
  start <- gets (tokenRange . curTkn)
  isNegative <- optional (tryConsumeExact (SymbolToken "-"))
  let sign = maybe 1 (const (-1)) isNegative
  num <- tryConsumeJust tryGetNumber

  end <- gets (tokenRange . prevTkn)
  pure (sign * num, rangeFromRanges start end)

-- | A parser for an expression. Returns Ignore if any of the sub expression parsers returns ignore,
--   fails if any of the sub expression parsers fails.
expressionParserM :: Parser' UncheckedExpression
expressionParserM = binaryParserM <|> intLiteralParserM <|> variableParserM
  where
    binaryParserM :: Parser' UncheckedExpression
    binaryParserM = do
      start <- gets (tokenRange . curTkn)
      -- Left side of a binary expression cannot be a recursive expression, to prevent infinite loops.
      left <- intLiteralParserM <|> variableParserM
      op <- operatorParserM
      right <- expressionParserM

      end <- gets (tokenRange . prevTkn)
      pure $ UBinary (rangeFromRanges start end) op left right

    intLiteralParserM :: Parser' UncheckedExpression
    intLiteralParserM = uncurry (flip UIntLiteral) <$> numberParserM

    variableParserM :: Parser' UncheckedExpression
    variableParserM = do
      start <- gets (tokenRange . curTkn)
      (var, varRange) <- tryConsumeJust' tryGetIdentifier
      exprMaybe <- optional expressionParserM

      end <- gets (tokenRange . prevTkn)
      pure $ UVariable (rangeFromRanges start end) (UncheckedVariable varRange var) exprMaybe

-- | A parser for a print statement. Returns Ignore if the print keyword cannot be parsed,
--   fails if anything later fails.
printStmtParserM :: Parser' (UncheckedStatement, Bool)
printStmtParserM = do
  start <- gets (tokenRange . curTkn)
  tryConsumeExact (KeywordToken "print")

  nextToken <- gets curTkn
  strLitAndRangeMaybe <- optional $ tryConsumeJust' tryGetStringLiteral
  exprMaybe <- optional expressionParserM

  end <- gets (tokenRange . prevTkn)
  let range = rangeFromRanges start end
  case (strLitAndRangeMaybe, exprMaybe) of
    (Just (sl, r), _) -> pure (UPrintStr range (UncheckedStringLiteral r sl), True)
    (_, Just e) -> pure (UPrintExpr range e, True)
    _ ->
      pfail $
        printf
          "%s Error parsing a print statement, expected a string literal or expression, but got %s."
          (fb $ tokenRange nextToken)
          (fb nextToken)

-- | A parser for anassignment statement. Returns Ignore if the let keyword cannot be parsed,
--   fails if anything later fails.
assignmentStmtParserM :: Natural -> Parser' (UncheckedStatement, Bool)
assignmentStmtParserM indent = do
  start <- gets (tokenRange . curTkn)
  tryConsumeExact (KeywordToken "let")
  (name, nameRange) <- consumeJust' tryGetIdentifier "assignment variable"
  paramAndRangeMaybe <- optional $ tryConsumeJust' tryGetIdentifier

  consumeExact (SymbolToken "=") "assignment"
  assmtRange <- gets (tokenRange . curTkn)
  end <- gets (tokenRange . prevTkn)

  (assmt, eols) <- assignmentParserM indent <|> pfail (printf "%s: Error parsing an assignment." (fb assmtRange))

  pure
    ( UAssignment
        (rangeFromRanges start end)
        (UncheckedVariable nameRange name)
        (uncurry (flip UncheckedVariable) <$> paramAndRangeMaybe)
        assmt,
      eols
    )

-- | A parser for a return statement. Returns Ignore if the return keyword cannot be parsed,
--   fails if anything later fails.
returnStmtParserM :: Parser' (UncheckedStatement, Bool)
returnStmtParserM = do
  start <- gets (tokenRange . curTkn)
  tryConsumeExact (KeywordToken "return")
  expr <- expressionParserM <|> pfail (printf "%s Error parsing a return expression" (fb start))

  end <- gets (tokenRange . prevTkn)
  pure (UReturn (rangeFromRanges start end) expr, True)

-- | A parser for a statement. Returns Ignore if any of the sub parsers returns ignore.
--   Fails when any of the sub parsers or the  end of line parser fails.
statementParserM :: Natural -> Parser' UncheckedStatement
statementParserM indent = do
  indentParserM indent
  (stmt, eols) <- printStmtParserM <|> assignmentStmtParserM indent <|> returnStmtParserM
  if eols then do eolsParser "statement" else pure ()
  pure stmt

-- | A parser for a program.
programParser :: Parser' UncheckedProgram
programParser = do
  start <- gets (tokenRange . curTkn)
  stmts <- many (statementParserM 0)
  end <- gets (tokenRange . prevTkn)
  pure $ UncheckedProgram (rangeFromRanges start end) stmts

-- | A parser for a project.
projectParser :: Parser' UncheckedProject
projectParser = do
  projectType <- projectTypeParser

  consumeIf isSeparator "separator"
  eolsParser "separator"

  program <- programParser

  consumeExact EndOfInputToken "project"
  pure $ UncheckedProject projectType program
